core/roboconf-dm-monitoring/src/main/java/net/roboconf/dm/monitoring/internal/TemplateWatcher.java

```diff
@@ -0,0 +1,360 @@
+		} finally {
+			this.lock.readLock().unlock();
+		}
+	}
+
+	/**
+	 * Compile the given template file and create the associated template entry.
+	 *
+	 * <p>IO and compile errors are logged but not rethrown.</p>
+	 *
+	 * @param templateFile the template file to compile.
+	 * @return the created template entry, or {@code null} if any problem has occurred.
+	 */
+	private TemplateEntry compileTemplate( final File templateFile ) {
+		TemplateEntry templateEntry = null;
+		try {
+			// Compile the template file.
+			final Template template = handlebars.compile(
+					new StringTemplateSource(
+							templateFile.toString(),
+							Utils.readFileContent( templateFile ) ) );
+			// Create the entry.
+			templateEntry = new TemplateEntry(
+					templateFile,
+					TemplateEntry.getTemplateId( templateFile ),
+					TemplateEntry.getTemplateApplicationName( templateDir, templateFile ),
+					template );
+		} catch (final IOException e) {
+			this.logger.warning( "Cannot compile template " + templateFile );
+			Utils.logException( this.logger, e );
+		} catch (final IllegalArgumentException e) {
+			this.logger.warning( "Cannot compile template " + templateFile );
+			Utils.logException( this.logger, e );
+		} catch (final HandlebarsException e) {
+			this.logger.warning( "Cannot compile template " + templateFile );
+			Utils.logException( this.logger, e );
+		}
+		return templateEntry;
+	}
+
+	/**
+	 * Add/update the given template entry in the {@code templates} map.
+	 *
+	 * @param templateEntry the template entry to add/update.
+	 * @GuardedBy this.lock.writeLock()
+	 */
+	private void updateTemplateEntry( final TemplateEntry templateEntry ) {
+		Map<String, TemplateEntry> specificTemplates = this.templates.get( templateEntry.appName );
+		if (specificTemplates == null) {
+			specificTemplates = new HashMap<String, TemplateEntry>();
+			this.templates.put( templateEntry.appName, specificTemplates );
+		}
+		specificTemplates.put( templateEntry.id, templateEntry );
+	}
+
+	//
+	// FileAlterationListener methods.
+	//
+
+	@Override
+	public void onStart( final FileAlterationObserver observer ) {
+		if (!this.hasBeenProvisioned.getAndSet( true )) {
+			// We must list the templates already present in the directory, as they won't otherwise trigger any event,
+			// and thus be ignored.
+			this.logger.fine( "Initial provisioning of templates..." );
+
+			// Find all the template files: the global *and* the specific ones.
+			final Collection<File> templateFiles = FileUtils.listFiles(
+					this.templateDir,
+					// Regular file filter:
+					// List readable files with the template extension.
+					FileFilterUtils.and(
+							FileFilterUtils.suffixFileFilter( MonitoringService.MONITORING_TEMPLATE_FILE_EXTENSION ),
+							CanReadFileFilter.CAN_READ ),
+					// Directory filter:
+					// Go through the root template directory and its direct children.
+					new AbstractFileFilter() {
+						@Override
```

Comment:
Le 27/04/2015 16:56, Vincent Zurczak a écrit :
>
> In 
> core/roboconf-dm-monitoring/src/main/java/net/roboconf/dm/monitoring/internal/TemplateWatcher.java 
> <https://github.com/roboconf/roboconf-platform/pull/295#discussion_r29153592>:
>
> > +			// We must list the templates already present in the directory, as they won't otherwise trigger any event,
> > +			// and thus be ignored.
> > +			this.logger.fine( "Initial provisioning of templates..." );
> > +
> > +			// Find all the template files: the global *and* the specific ones.
> > +			final Collection<File> templateFiles = FileUtils.listFiles(
> > +					this.templateDir,
> > +					// Regular file filter:
> > +					// List readable files with the template extension.
> > +					FileFilterUtils.and(
> > +							FileFilterUtils.suffixFileFilter( MonitoringService.MONITORING_TEMPLATE_FILE_EXTENSION ),
> > +							CanReadFileFilter.CAN_READ ),
> > +					// Directory filter:
> > +					// Go through the root template directory and its direct children.
> > +					new AbstractFileFilter() {
> > +						@Override
>
> It is not about code clarity but about potential memory leaks.
>
In fact, the inner class keeps a pointer to the parent class instance.

So, the problem is: even if there is no visible pointer to the instance 
of the parent class, it will not be garbage-collected... as long as an 
instance of the inner class exists somewhere.

For example, if you put the inner class instance in a Hashmap, or keep 
it as another instance's field value, then the parent instance may 
remain forever.

So, using anonymous inner classes is *potentially* dangerous (as it is 
error prone...), although it *works* (not a JVM bug nor limitation)...